relay-cardinality/src/cache.rs

```diff
@@ -0,0 +1,251 @@
+        window: u64,
+        limit: usize,
+    ) -> CacheRead<'_> {
+        let inner = self.inner.read().unwrap_or_else(PoisonError::into_inner);
+
+        // If the window does not match, we can already release the lock.
+        if window != inner.current_window {
+            return CacheRead::noop();
+        }
+
+        CacheRead::new(inner, organization_id, limit)
+    }
+
+    /// Acquires a write lock from the cache and returns an update handle.
+    ///
+    /// All operations done on the handle share the same lock. To release the lock
+    /// the returned [`CacheUpdate`] must be dropped.
+    pub fn update<'a>(
+        &'a self,
+        organization_id: OrganizationId,
+        item_scope: &'a str,
+        window: u64,
+    ) -> CacheUpdate<'a> {
+        let mut inner = self.inner.write().unwrap_or_else(PoisonError::into_inner);
+
+        // If the window is older don't do anything and give up the lock early.
+        if window < inner.current_window {
+            return CacheUpdate::noop();
+        }
+
+        // If the window is newer then the current window, reset the cache to the new window.
+        if window > inner.current_window {
+            inner.current_window = window;
+            inner.cache.clear();
+        }
+
+        let key = CacheKeyRef {
+            organization_id,
+            item_scope,
+        };
+        CacheUpdate::new(inner, key)
+    }
+}
+
+/// Cache read handle.
+///
+/// Holds a cache read lock, the lock is released on drop.
+pub struct CacheRead<'a>(CacheReadInner<'a>);
+
+/// Internal state for [`CacheRead`].
+enum CacheReadInner<'a> {
+    Noop,
+    Cache {
+        inner: RwLockReadGuard<'a, Inner>,
+        organization_id: OrganizationId,
+        limit: usize,
+    },
+}
+
+impl<'a> CacheRead<'a> {
+    /// Creates a new [`CacheRead`] which reads from the cache.
+    fn new(
+        inner: RwLockReadGuard<'a, Inner>,
+        organization_id: OrganizationId,
+        limit: usize,
+    ) -> Self {
+        Self(CacheReadInner::Cache {
+            inner,
+            organization_id,
+            limit,
+        })
+    }
+
+    /// Creates a new noop [`CacheRead`] which does not require a lock.
+    fn noop() -> Self {
+        Self(CacheReadInner::Noop)
+    }
+
+    pub fn check(&self, item_scope: &str, hash: u32) -> CacheOutcome {
+        match &self.0 {
+            CacheReadInner::Noop => CacheOutcome::Unknown,
+            CacheReadInner::Cache {
+                inner,
+                organization_id,
+                limit,
+            } => inner
+                .cache
+                .get(&CacheKeyRef {
+                    organization_id: *organization_id,
+                    item_scope,
+                })
+                .map_or(CacheOutcome::Unknown, |s| s.check(hash, *limit)),
+        }
+    }
+}
+
+/// Cache update handle.
+///
+/// Holds a cache write lock, the lock is released on drop.
+pub struct CacheUpdate<'a>(CacheUpdateInner<'a>);
```

Comment:
Wrapper struct is new, `CacheUpdateInner` can't be public without making `Inner` public, which is not what we want.